### ISSUE TITLE
f32 to u32 cast has different result in different context

### ISSUE THREAD

**Issue by bluss** [rust-lang/rust#56782]
A `f32 as u32` cast has different result in two different places.

This [code (playground link):](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=42e34941c171722708c5fb380bce0149)

```rust
fn main() {
    println!("{:?}", {
        u32::max_value() as f32 as u32
    });
    let v = u32::max_value() as f32 as u32;
    println!("{:?}", v);
}
```

has the output:

```
4294967295
0
```

which is a surprise! We expect the output should be the same in both cases. Even if the temporary f32 value is outside of the range of values exactly representable in u32.

Rustc version: 1.31.0

**Comment by bluss** [comment]
Basically #10184

**Comment by rkruppe** [comment]
Yes, ultimately due to #10184. The immediate reason there's different output is that `println` takes a reference to its argument so it triggers promotion, and CTFE handles these casts differently (saturating rather than returning `undef`) than runtime code generation w/o `-Zsaturating-float-casts`.

**Comment by rkruppe** [comment]
@bluss Can this be closed as duplicate of #10184?

**Comment by bluss** [comment]
Ok
